File: leetcode/medium/Problem29.java
package leetcode.medium;


/**
 * Leetcode 29
 * <p>
 * Link: <a href="https://leetcode.com/problems/divide-two-integers/description/">Problem</a>
 * <p>
 * Hint: Utilize bitwise operations to find the largest multiply first and then subtract for better performance
 * */

public class Problem29 {
    public static int divide(int dividend, int divisor) {
        if (dividend == 0) return 0;
        if (dividend == Integer.MIN_VALUE && divisor == -1) return Integer.MAX_VALUE;
        if (dividend == Integer.MIN_VALUE && divisor == 1) return Integer.MIN_VALUE;

        boolean negative = (dividend < 0) ^ (divisor < 0);
        long absDividend = Math.abs((long) dividend);
        long absDivisor = Math.abs((long) divisor);

        int res = 0;
        while (absDividend >= absDivisor) {
            long tempDivisor = absDivisor;
            long multiply = 1;
            // Find the largest multiply by shifting for better performance
            while (absDividend >= (tempDivisor << 1)) {
                tempDivisor <<= 1;
                multiply <<= 1;
            }
            absDividend -= tempDivisor;
            res += multiply;
        }
        return negative ? -res : res;
    }

    public static void main(String[] args) {
        System.out.println(divide(10, 3));
        System.out.println(divide(7, -3));
        System.out.println(divide(-99, -2));
        System.out.println(divide(99, -7));
        System.out.println(divide(-2147483648, -1));
        System.out.println(divide(-2147483648, 2));
        System.out.println(divide(1000000000, 1));
    }
}

File: leetcode/python/medium/Problem29.py
# Questions to ask:
# 1. What is the time complexity?
# 2. What is the space complexity?
class Solution:
    def divide(self, dividend: int, divisor: int) -> int:
        MIN_NUM = -2147483648
        MAX_NUM = 2147483647

        negative = (dividend < 0) ^ (divisor < 0)
        dividend = abs(dividend)
        divisor = abs(divisor)
        quotient = dividend // divisor
        return min(MAX_NUM, max(MIN_NUM, quotient if not negative else -quotient))


    def divide2(self, dividend: int, divisor: int) -> int:
        MIN_NUM = -2147483648
        MAX_NUM = 2147483647

        quotient = 0
        negative = (dividend < 0) ^ (divisor < 0)
        dividend = abs(dividend)
        divisor = abs(divisor)

        while dividend >= divisor:
            temp = divisor
            multiply = 1
            # Find the largest multiply by shifting for better performance
            while dividend >= (temp << 1):
                temp <<= 1
                multiply <<= 1
            dividend -= temp
            quotient += multiply

        return min(MAX_NUM, max(MIN_NUM, quotient if not negative else -quotient))

# Problem 22
# Link: https://leetcode.com/problems/divide-two-integers/description/
if __name__ == '__main__':
    s = Solution()
    cases = [
        (10, 3, 3),
        (7, -3, -2),
    ]
    for dividend, divisor, expected in cases:
        assert s.divide(dividend, divisor) == expected
        assert s.divide2(dividend, divisor) == expected
